service/storage_service.cc

```diff
@@ -1835,15 +1840,104 @@ future<> storage_service::decommission() {
             }
 
             slogger.info("DECOMMISSIONING: starts");
-            ss.start_leaving().get();
-            // FIXME: long timeout = Math.max(RING_DELAY, BatchlogManager.instance.getBatchlogTimeout());
-            auto timeout = ss.get_ring_delay();
-            ss.set_mode(mode::LEAVING, format("sleeping {} ms for batch processing and pending range setup", timeout.count()), true);
-            sleep_abortable(timeout, ss._abort_source).get();
+            auto leaving_nodes = std::list<gms::inet_address>{endpoint};
+            std::list<gms::inet_address> ignore_nodes;
+
+            future<> heartbeat_updater = make_ready_future<>();
+            auto heartbeat_updater_done = make_lw_shared<bool>(false);
+
+            // Step 1: Decide who needs to sync data
+            std::vector<gms::inet_address> nodes;
+            for (const auto& x : tmptr->get_endpoint_to_host_id_map_for_reading()) {
+                seastar::thread::maybe_yield();
+                if (std::find(ignore_nodes.begin(), ignore_nodes.end(), x.first) == ignore_nodes.end()) {
+                    nodes.push_back(x.first);
+                }
```

Comment:
Wouldn't a `std::unordered_set<gms::inet_address> ignore_nodes` be more efficient?

Reply:
Hmm, wait, `ignore_nodes` is empty at this time.
Worth a FIXME comment as you added below in `run_bootstrap_ops`

Reply:
To avoid big allocations if the list is large. We use list for the data we send over network. Keep it the same to avoid convert back and forth.

I added a TODO for dedecommission. We already have a TODO comment for run_bootstrap_ops.